### DISCUSSION TITLE
Why did we choose Google Compute (VPS) instead of Google Run (PAAS)?

### DISCUSSION THREAD

**Discussion by gth001** [NightscoutFoundation/xDrip#2492] [General]
Sorry if this is a dumb question, but I am just curious! ~~Someone on facebook mentioned that [Railway is easier to secure than Google](https://www.facebook.com/groups/cgminthecloud/posts/5686285998159276/). They said it's because one is PAAS (you manage the app) and one is VPS (you manage the server and app). I wanted to learn about the difference, and notice that Google actually offers both.~~ (deleted as irrelevant to the question) How come we are we using https://cloud.google.com/compute instead of https://cloud.google.com/run ? It says "All customers get 2 million requests free per month" - so is that too little for nightscout? It sounds like it would eliminate a lot of steps from [the instructions](https://navid200.github.io/xDrip/docs/Nightscout/GoogleCloud.html). 

Thanks
-from a complete noobie

**Comment by Navid200** [comment]
> Sorry if this is a dumb question, but I am just curious! Someone on facebook mentioned that [Railway is easier to secure than Google](https://www.facebook.com/groups/cgminthecloud/posts/5686285998159276/). 

I am not sure if that's what was said.  But, it's quite possible that you misunderstood that post.

There is no security issue with our setup.
If you want it to be free, you need to limit the outgoing traffic.  That is one thing that we didn't have to worry about with Heroku.
Because of that, until we have evidence, we are suggesting to everyone to not open their site to the world as they may have done with Heroku.
This is not because we believe that Heroku was more secure than our setup.  
We do that because we want to limit the traffic to ensure it will be free.  This has nothing to do with security and everything to do with cost.

We can go into more detail if needed.  
But, until we agree on this one way or another, it is pointless to proceed with the remaining parts of your post.

**Comment by gth001** [comment]
I'll delete the part of question about security. I was more curious why we are managing a virtual server when Google offers to manage it for free.

The reason is that so many users are overwhelmed by the number of steps in the instructions, to the point that they're paying Heroku money. I just thought it'd be nice to eliminate half of the steps. I guess I could give it a shot, I just figured someone else probably tried already and can tell me why it's a bad idea.

**Reply by Navid200** [reply]
People complain that the instructions are not easy to follow.
So, I break it down into more steps, and add lots of other links to cover what may go wrong and how to deal with it in each case, and add screenshots and video clips, and then, it is too many steps!
There are people who tell me now that the instructions are not easy to follow.  They don't say it is too many steps.  They say they are not sure what they should copy and paste and I should change the font to be larger.

My point is that there is absolutely no way to make everyone happy.  There will always be someone who would prefer something else.  And that is fine.

Anyway, where would the MongoDB be stored if you use Cloud run?

How easy would it be for a user to customize Nightscout on Cloud Run?

When the next Nightscout version is released, we will release a new version of our fork so that users can run the install steps again to update.  But, they don't have to wait for us.  They can just update their Nightscout fork on GitHub and customize Nightscout using the menu item for it and install from their fork and done; they will be running the latest version of Nightscout even before we release our update.
Will that be possible in a container?

When we find a bug, we release an update and all the user needs to do is to select Update scripts from the menu to update all their executables to the latest release.  How will that work in a container?
I mean we don't want the user to have to start from scratch for every new update.  All of our releases are incremental.

We are still learning.  I know for sure I have a lot to learn.
I didn't know anything about Google compute engine 4 months ago.  I still know nothing about Google cloud run.

**Reply by gth001** [reply]
>and then, it is too many steps!

Not at all what I meant. You're doing a great job Navid. Thank you for your work. The increased number of steps is not your doing, it's entirely due to the choice of Google Compute instead of Run. Maybe @jamorham knows the reason.

>When the next Nightscout version is released, we will release a new version of our fork so that users can run the install steps again to update. But, they don't have to wait for us. They can just update their Nightscout fork on GitHub and customize Nightscout using the menu item for it and install from their fork and done; they will be running the latest version of Nightscout even before we release our update.
Will that be possible in a container?

Railway is PAAS and my impression is that a lot of CGM in the Cloud users are choosing it. I assume it can update and customize the same way as you're describing. Azure and Heroku are PAAS too. If they are using containers, I don't know, that stuff is invisible to the user. I am just wondering why the Google instructions are unique in rejecting PAAS. 

I feel like there's a very simple answer to my question out there, and I came here totally expecting a one sentence reply where I would then say "oh that makes total sense, thanks". I am not criticizing or nitpicking here. If my question was a bother, I'll just forget about it. Thank you all so much! This stuff and all your efforts make my daily life great, and especially you Navid have fixed so many issues for me over the years. Anyway have a good one.

**Reply by Navid200** [reply]
JamOrHam came up with this idea and it has all been new to me.
So, I did not mean to offer an answer to your question.  I don't have an answer.
Your question has not been a bother.

**Comment by jamorham** [comment]
@gth001 google cloud run doesn't include as much cpu in the free tier afaik and so wouldn't work for always on features. With running things as a service PAAS there can be issues, for example people would post that their xDrip would stop uploading to nightscout when actually it was the database being over limits or laggy on the service provider. With a vps this is also directly under our control and should be more stable as nothing changes unless you want it to. The original script I put together was designed to be a single paste in and the only user input required was the credentials for the free dns and certificate entry. After that it should require no maintenance at all - ever... I put that script together considering my own use really as my needs are fairly simple but it has been extended by @Navid200 and @tzachi-dar to include many more features such as data loading, menus, ability to upgrade version, use custom versions, change variables etc. I expect additional complexity has been introduced there but that is of course likely with additional features. I think the main thing is that users have choices about how and where they store and manage their data and things don't go dark because of changes at a single service provider.

**Reply by gth001** [reply]
Oh ok, that makes sense, thank you all
